[PT2E][Quant] Use subgraph matcher annotate linear pattern

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99580
* #99399
* #99398

This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @soumith @desertfire

kimishpatel committed:
[PT2E][Quant] Use subgraph matcher annotate linear pattern

This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
+                    bias_node = ph
+
+            # bias and output act
+            if _is_annotated([act_node]) == False:
```
nit: `not _is_annotated([act_node])`?

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
+                    "output_act_obs_or_fq_ctr": get_act_obs_or_fq_ctr(
+                        quantization_config
+                    ),
+                    "weight_obs_or_fq_ctr": None,
+                    "bias_obs_or_fq_ctr": None,
```
these are optional I think, you can remove them

jerryzh168 reviewed (approved):
LGTM, should be able to cleanup the annotation further after https://github.com/pytorch/pytorch/pull/99220 is landed

leslie-fang-intel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
+            # bias and output act
+            if _is_annotated([act_node]) == False:
+                act_node.meta["target_dtype_info"] = {
+                    "input_act_obs_or_fq_ctr": None,
```
May I ask a question why `input_act_obs_or_fq_ctr` is None here for `act_node`? Does it mean we don't insert observer for the input of  `act_node`?

jerryzh168 replied:
`act_node` here is actually the node that produces the activation for linear, so the output of `act_node` is the input for linear

kimishpatel commented:
> LGTM, should be able to cleanup the annotation further after #99220 is landed

Yes, that would be good, but lets catch up on what that means. I explicitly annotated in between nodes without anything,  because I expected quant workflow to not touch those nodes. But I want to sync with you whether thats a valid assumption.

jerryzh168 commented:
> > LGTM, should be able to cleanup the annotation further after #99220 is landed
> 
> Yes, that would be good, but lets catch up on what that means. I explicitly annotated in between nodes without anything, because I expected quant workflow to not touch those nodes. But I want to sync with you whether thats a valid assumption.

yeah you can annotate with None to mean don't touch this, after my PR you also don't need to specify anything and it will just work (assumed to be torch.float after that PR)

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Use subgraph matcher annotate linear pattern"


This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
